### PR TITLE
Dynamic Gneiss on Linux

### PR DESCRIPTION
**Pull request by jklmnn**
This PR implements #119, #126 and #130. It is in a proof of concept state and things might break in complex or large scenarios. I still think we should merge it now since the pull request will eventually grow even larger.
It can be built with cement:
```
$ ./cement <xml config> <root of gneiss repo> [include paths ...]
```
e.g. the message example can be run with
```
$ ./cement test/message_client/message_client.xml . init test lib
$ export LD_LIBRARY_PATH=build/lib
$ ./build/bin/init test/message_client/message_client.xml
```

As a proof of concept [Ceres](https://github.com/jklmnn/ceres) provides a bare metal Linux distribution based on Gneiss. It can be tested by cloning it recursively and running
```
./build.sh gneiss/test/message_client/message_client.xml
```
It expects a Debian 10 system and GNAT Community 2019.

### CODE REVIEW

**Comment by senier** [comment]
Very nice!

I just tested it and it seems to work. I'm wondering about one warning message, through:

```sh
I: <config>
    <component name="message_client" file="libcomponent_message_client.so">
        <service name="Message" server="message_server"/>
    </component>    <component name="message_server" file="libcomponent_message_server.so">
    </component>
</config>
I: Started message_client with PID 14
I: Started message_server with PID 15
[message_client:log] Hello
W: Invalid message, dropping
I: Component message_client exited with status 0
```

Do you have an idea where that's coming from?

**Comment by jklmnn** [comment]
The warning message is usually printed when a packet is received that could not be parsed. This is probably the EOF the socket receives when the other side closes it. I plan to remove the message in the future I's currently a debug output that helps in case of missing messages (e.g. an expected message is missing this output would indicate that it arrived and could not be parsed).

File: init/gneiss-broker.adb [hunk 1]
```diff
@@ -0,0 +1,771 @@
+
+with Basalt.Strings;
+with Gneiss.Protocoll;
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply init/gneiss-broker.adb]: Fixed by https://github.com/Componolit/gneiss/pull/131/commits/1253cad39d2138529d24093af720df7c767a369a.

File: test/message_client.gpr [hunk 2]
```diff
@@ -0,0 +1,13 @@
+
+library project Message_Client extends "gneiss_component" is
+
+   for Runtime ("Ada") use Gneiss_Runtime;
+   for Source_Dirs use ("message_client");
+   for Object_Dir use external ("CEMENT_OBJECT_DIR");
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply test/message_client.gpr]: I tried but `gneiss_component.gpr` is already the maximum I can move to a generic file.

File: init/gneiss-broker.adb [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+      Gneiss_Log.Info (Config);
+      Status := 1;
+      Policy := (others => (Fd => -1, Node => SXML.Query.Init (Document => Document), Pid => -1));
+      if not SXML.Valid_Content (Config'First, Config'Last) then
+         Gneiss_Log.Error ("Invalid content");
+         return;
+      end if;
+      SXML.Parser.Parse (Config, Document, Ignore_Pos, Result);
+      if
+         Result /= SXML.Parser.Match_OK
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply init/gneiss-broker.adb]: Fixed by https://github.com/Componolit/gneiss/pull/131/commits/667f25e75bddcb1070d3bc922a47d40a0dffb8a4.

File: lib/RecordFlux/src/rflx-lemmas.adb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+package body RFLX.Lemmas with
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply lib/RecordFlux/src/rflx-lemmas.adb]: Fixed by https://github.com/Componolit/gneiss/pull/131/commits/c9be183d09873c654edce97a5495a7ae62ecde08.

**Pull request merged**